Refactor Todo settings, offer pre-defined list of Todo apps

yourcontact opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

### Description
<!--- Describe your changes in detail -->
As per #418, in `Todo Server` setting, I substituted the field where users had to enter the url of the todo app they use with a dropdown menu displaying a list of todo apps:

- [Todoist](https://todoist.com/)
- Franz Todo
- [TickTick](https://www.ticktick.com/)
- Microsoft To Do
- [Habitica](https://habitica.com)
- [Nozbe](https://nozbe.com/)
- [Remember The Milk](https://www.rememberthemilk.com)
- [Any.do](https://www.any.do/ ) (the log in page isn't responsing in Ferdi todo's webview)

Please let me know if you want to delete any of the above, or add any.

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
The change adds convenience for users.
Now, users can select the todo app they use instead of fetching and copying/pasting the url of the todo app they use.

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->
I tested in Electron. I am not sure how to test it in as a new user cc @kytwb 

### Screenshots (if appropriate):

![Kapture 2020-03-24 at 22 38 11](https://user-images.githubusercontent.com/46404814/77479629-70480b80-6e20-11ea-8348-8305b2e3319c.gif)

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

That is a refactoring.

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project (run `$ yarn lint`). 
<!---- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. -->

@kytwb @vantezzen does my refactoring require a change to the documentation?

kytwb commented:
Thanks for picking this one up @yourcontact 💪
Looking forward to the rest and happy to assist again whenever 🙏

vantezzen commented:
Looks great.

A suggestion would be to add another option "Other service" that displays a text input box below the `select`-element so users can use any service they want.
This would need to be implemented similar to how the darkmode setting displays more sub-settings when you activate it ([Component](https://github.com/getferdi/ferdi/blob/develop/src/components/settings/settings/EditSettingsForm.js#L462-L476), [Container](https://github.com/getferdi/ferdi/blob/develop/src/containers/settings/EditSettingsScreen.js#L548)).
This would also require a change to [the webview](https://github.com/getferdi/ferdi/blob/develop/src/features/todos/components/TodosWebview.js#L226) to use that custom value istead if selected.

yourcontact commented:
Thank you @vantezzen.

I would like to share an update on where I am with the refactoring Ferdi Todo setting:

- added "Other service" to the drop down menu
- added a field that appears when "Otherservice" is selected in the dropdown menu

Next step is to fix the field for users to input URL of the todo service/app they use. The field currently collapse when I try to input data.

![HptD0U](https://user-images.githubusercontent.com/46404814/77162560-87979980-6aac-11ea-8ad3-fd5734dd1261.gif)

vantezzen commented on src/components/settings/settings/EditSettingsForm.js:
```diff
@@ -34,7 +34,7 @@ const messages = defineMessages({
   headlineGeneral: {
     id: 'settings.app.headlineGeneral',
-    defaultMessage: '!!!General',
+    defaultMessage: '!!!Generadddl',
```
Revert this change

yourcontact replied:
I fixed that typo https://github.com/yourcontact/ferdi/commit/62f1a167bb56a5f62a56bcd88fd67f6341e9dc32

vantezzen commented on src/components/settings/settings/EditSettingsForm.js:
```diff
@@ -333,28 +333,37 @@ export default @observer class EditSettingsForm extends Component {
+                {form.$('enableTodos').value && (
+                  <div>
+                    <Select field={form.$('todoServer')} />
+                    {form.$('todoServer').value === 'otherService' && (
```
You should do this in the container element (src/containers/settings/EditSettingsScreen) and pass something like `idTodosActivated` and `isUsingCustomTodoService` prop to the component instead - at least this is how we've done it for most other settings like this.

vantezzen commented on src/i18n/locales/af.json:
```diff
@@ -272,7 +272,7 @@
   "settings.app.form.adaptableDarkMode": "Synchronize dark mode with my Mac's dark mode setting",
   "settings.app.form.autoLaunchInBackground": "Open in background",
-  "settings.app.form.autoLaunchOnStart": "Launch Ferdi on start",
+  "settings.app.form.autoLaunchOnStart": "Launch sdfsdf Ferdi on start",
```
Revert that change

vantezzen commented on src/i18n/locales/en-US.json:
```diff
@@ -302,7 +302,7 @@
   "settings.app.form.showMessagesBadgesWhenMuted": "Show unread message badge when notifications are disabled",
   "settings.app.form.startMinimized": "Start minimized in tray",
-  "settings.app.form.todoServer": "Todo Server",
+  "settings.app.form.todoServer": "Todo Service",
```
You should also change this in https://github.com/getferdi/ferdi/blob/cecdbc4689/src/containers/settings/EditSettingsScreen.js#L85

yourcontact replied:
I reverted that change

yourcontact commented:
@vantezzen I resolved your comments https://github.com/getferdi/ferdi/pull/477#pullrequestreview-378435525 and updated my initial PR comment https://github.com/getferdi/ferdi/pull/477#issue-390407136 with latest change. 

I am not sure how to fix this (cc @kytwb):
<img width="671" alt="Screen Shot 2020-03-24 at 10 48 35 PM" src="https://user-images.githubusercontent.com/46404814/77480284-b0f45480-6e21-11ea-95d6-13648cc4b0ba.png">

kytwb commented:
@yourcontact can you try removing these files?

<img width="797" alt="Screen Shot 2020-03-25 at 12 43 28 AM" src="https://user-images.githubusercontent.com/412895/77487055-ab9f0600-6e31-11ea-80d7-b2c2d5e9c639.png">

kytwb commented:
@yourcontact feel free to post updates in new comments instead of editing past comments, it's less confusing for the thread history 🙏 no big deal though

vantezzen commented:
We will still need a migration from the old settings to the new ones in https://github.com/getferdi/ferdi/blob/develop/src/stores/SettingsStore.js#L185. I had a custom Todo server set but the settings screen says that I selected Todoist:

![Screenshot 2020-03-25 at 10 42 49](https://user-images.githubusercontent.com/10333196/77522990-60aede00-6e85-11ea-8302-f9949c619945.png)

kytwb commented:
@vantezzen Good catch! 
Regarding the errors, we should probably try to load the Custom Todo URL only once it's a valid URL.

yourcontact commented:
> @yourcontact can you try removing these files?
> 
> <img alt="Screen Shot 2020-03-25 at 12 43 28 AM" width="797" src="https://user-images.githubusercontent.com/412895/77487055-ab9f0600-6e31-11ea-80d7-b2c2d5e9c639.png">

I removed those files.

yourcontact commented:
@kytwb and I added a regex to validate the URL when a user select `Other service`.

>We will still need a migration from the old settings to the new ones in https://github.com/getferdi/ferdi/blob/develop/src/stores/SettingsStore.js#L185.

@kytwb will take over this pull request and set a migration from old settings.

yourcontact commented:
@vantezzen I will work on the migration of old settings with @kytwb tomorrow.

kytwb commented:
I had to get going with my day, so we only worked on https://github.com/getferdi/ferdi/pull/506; we're pairing again tomorrow at 9am (my local time) to address the missing migration.

yourcontact commented:
I resolved conflicts in src/containers/settings/EditSettingsScreen.js.

For the conflicts in the two following files:
- src/i18n/locales/defaultMessages.json;
- src/i18n/messages/src/containers/settings/EditSettingsScreen.json;

It looks like there are conflicts due to code style changes. 

Shall I apply what is in the "refactor/settings-todo" branch and delete what is on the "develop" branch?

kytwb commented:
I am not sure myself what's the best thing to do right now. Can you try keeping your changes and re-running `npm run manage-translations`?

yourcontact commented:
> I am not sure myself what's the best thing to do right now. Can you try keeping your changes and re-running npm run manage-translations?

@kytwb I ran `npm run manage-translations` on the "refactor/settings-todo" branch.
Below is a screenshot of what I got. I am not sure what to do next.
<img width="570" alt="Screen Shot 2020-03-31 at 12 08 06 PM" src="https://user-images.githubusercontent.com/46404814/78014576-6c4c4a00-7348-11ea-9eb0-ec6a883174e8.png">

kytwb commented:
Check if the diff looks good to you (`git diff`), then you can finalize the merge with `git commit --no-edit`.

kytwb commented:
Not sure what's the best way to handle this given the restrictions enforced (scenario where we need to help in pull requests from external contributors), but I think it's best for the long term if we figure it out.

kytwb commented:
Closed by #527.
